src/m365/entra/commands/app/app-permission-list.ts

```diff
@@ -0,0 +1,248 @@
+      requiredResourceAccess.map(resourceAccess => {
+        return {
+          appId: resourceAccess.resourceAppId as string
+        };
+      });
+    const servicePrincipals = await Promise
+      .all(servicePrincipalsToResolve.map(servicePrincipalInfo =>
+        this.getServicePrincipal(servicePrincipalInfo, permissionType, logger) as ServicePrincipal));
+
+    const apiPermissions: ApiPermission[] = [];
+    requiredResourceAccess.forEach(requiredResourceAccess => {
+      const servicePrincipal = servicePrincipals
+        .find(servicePrincipal => servicePrincipal?.appId === requiredResourceAccess.resourceAppId as string);
+      const resourceName = servicePrincipal?.displayName as string ?? requiredResourceAccess.resourceAppId as string;
+      (requiredResourceAccess.resourceAccess as ResourceAccess[]).forEach(permission => {
+        if (permissionType === 'application' && permission.type === 'Scope') { return; }
+        if (permissionType === 'delegated' && permission.type === 'Role') { return; }
+
+        apiPermissions.push({
+          resource: resourceName,
+          resourceId: requiredResourceAccess.resourceAppId,
+          permission: this.getPermissionName(permission.id as string, permission.type as string, servicePrincipal),
+          type: permission.type === 'Role' ? 'Application' : 'Delegated'
+        });
+      });
+    });
+
+    return apiPermissions;
+  }
+
+  private getPermissionName(permissionId: string, permissionType: string, servicePrincipal: ServicePrincipal | undefined): string {
+    if (!servicePrincipal) {
+      return permissionId;
+    }
+
+    if (permissionType === 'Role') {
+      return (servicePrincipal.appRoles as AppRole[])
+        .find(appRole => appRole.id === permissionId)?.value as string ?? permissionId;
+    }
+    else {
+      // permissionType === 'Scope'
+      return (servicePrincipal.oauth2PermissionScopes as PermissionScope[])
+        .find(permissionScope => permissionScope.id === permissionId)?.value as string ?? permissionId;
+    }
+  }
+
+  private async getServicePrincipal(servicePrincipalInfo: ServicePrincipalInfo, permissionType: string, logger: Logger): Promise<ServicePrincipal | null> {
+    if (this.verbose) {
+      await logger.logToStderr(`Retrieving service principal ${servicePrincipalInfo.appId}`);
+    }
+
+    const requestOptions: CliRequestOptions = {
+      url: `${this.resource}/v1.0/servicePrincipals?$filter=appId eq '${servicePrincipalInfo.appId}'&$select=appId,id,displayName`,
+      headers: {
+        accept: 'application/json;odata.metadata=none'
+      },
+      responseType: 'json'
+    };
+
+    const response = await request.get<{ value: ServicePrincipal[] } | ServicePrincipal>(requestOptions);
+
+    if (servicePrincipalInfo.appId && (response as { value: ServicePrincipal[] }).value.length === 0) {
+      return null;
+    }
+
+    const servicePrincipal = (response as { value: ServicePrincipal[] }).value[0];
```

Comment:
just noticed it doing the final glance. If we assume the possibility to return ServicePrincipal won't this actually fail? We assume in line 207 that it will be an array with value property but if the request returns `ServicePrincipal` it will actually break right?

Reply:
Good catch! The return type as ServicePrincipal is a remnant of the code taken from `m365 app permission list` where a service principal could be requested according to the id or a filter. Since we only use a filter here, the response type will always be of the format `{value: ServicePrincipal[]}`